File: yogi-cpp/test/yogi/timestamp_test.cc
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <test/common.h>

using yogi::ArithmeticException;
using yogi::Duration;
using yogi::Timestamp;
using namespace std::chrono_literals;

class TimestampTest : public Test {};

TEST_F(TimestampTest, FromDurationSinceEpoch) {
  EXPECT_EQ(Timestamp::from_duration_since_epoch(123ns).duration_since_epoch(), 123ns);
  EXPECT_THROW(Timestamp::from_duration_since_epoch(Duration::kInf), ArithmeticException);
  EXPECT_THROW(Timestamp::from_duration_since_epoch(yogi::Duration(-1ns)), ArithmeticException);
}

TEST_F(TimestampTest, Now) {
  MOCK_GetCurrentTime([](long long* timestamp) {
    EXPECT_NE(timestamp, nullptr);
    *timestamp = 5555555555ll;
    return YOGI_OK;
  });

  EXPECT_EQ(Timestamp::now().duration_since_epoch().nanoseconds_count(), 5555555555ll);
}

TEST_F(TimestampTest, NowError) {
  MOCK_GetCurrentTime([](long long*) { return YOGI_ERR_UNKNOWN; });

  EXPECT_THROW(Timestamp::now(), FailureException);
}

TEST_F(TimestampTest, Parse) {
  MOCK_ParseTime([](long long* timestamp, const char* str, const char* timefmt) {
    EXPECT_NE(timestamp, nullptr);
    EXPECT_STREQ(str, "foo");
    EXPECT_EQ(timefmt, nullptr);
    *timestamp = 1234356789123000000ll;
    return YOGI_OK;
  });

  auto t = Timestamp::parse("foo");
  EXPECT_EQ(t.duration_since_epoch(), 1234356789123000000ns);
}

TEST_F(TimestampTest, ParseError) {
  MOCK_ParseTime([](long long*, const char*, const char* timefmt) {
    EXPECT_STREQ(timefmt, "bar");
    return YOGI_ERR_UNKNOWN;
  });

  EXPECT_THROW(Timestamp::parse("foo", "bar"), yogi::FailureException);
}

TEST_F(TimestampTest, Constructor) {
  EXPECT_EQ(Timestamp().duration_since_epoch(), 0ns);
}

TEST_F(TimestampTest, DurationSinceEpoch) {
  EXPECT_EQ(Timestamp::from_duration_since_epoch(123ns).duration_since_epoch(), 123ns);
}

TEST_F(TimestampTest, Nanoseconds) {
  EXPECT_EQ(Timestamp::from_duration_since_epoch(123456789ns).nanoseconds(), 789);
}

TEST_F(TimestampTest, Microseconds) {
  EXPECT_EQ(Timestamp::from_duration_since_epoch(123456789ns).microseconds(), 456);
}

TEST_F(TimestampTest, Milliseconds) {
  EXPECT_EQ(Timestamp::from_duration_since_epoch(123456789ns).milliseconds(), 123);
}

TEST_F(TimestampTest, Format) {
  auto t = Timestamp::from_duration_since_epoch(123456789123456789ns);

  MOCK_FormatTime([](long long timestamp, const char** str, int* strsize, const char* timefmt) {
    EXPECT_EQ(timestamp, 123456789123456789ll);
    EXPECT_NE(str, nullptr);
    EXPECT_NE(strsize, nullptr);
    EXPECT_EQ(timefmt, nullptr);
    *str     = "x";
    *strsize = static_cast<int>(strlen(*str) + 1);
    return YOGI_OK;
  });

  EXPECT_EQ(t.format(), "x");
}

TEST_F(TimestampTest, FormatError) {
  auto t = Timestamp::from_duration_since_epoch(123456789123456789ns);

  MOCK_FormatTime([](long long, const char**, int*, const char* timefmt) {
    EXPECT_STREQ(timefmt, "foo");
    return YOGI_ERR_UNKNOWN;
  });

  EXPECT_THROW(t.format("foo"), FailureException);
}

TEST_F(TimestampTest, ToString) {
  auto t = Timestamp::from_duration_since_epoch(123456789123456789ns);

  MOCK_FormatTime([](long long timestamp, const char** str, int* strsize, const char* timefmt) {
    EXPECT_EQ(timestamp, 123456789123456789ll);
    EXPECT_NE(str, nullptr);
    EXPECT_NE(strsize, nullptr);
    EXPECT_EQ(timefmt, nullptr);
    *str     = "x";
    *strsize = static_cast<int>(strlen(*str) + 1);
    return YOGI_OK;
  });

  EXPECT_EQ(t.to_string(), "x");
}

TEST_F(TimestampTest, Operators) {
  auto t = Timestamp();
  t += 8ns;
  EXPECT_EQ(t.duration_since_epoch(), 8ns);
  EXPECT_THROW(t += -1ms, ArithmeticException);
  t -= 1ns;
  EXPECT_EQ(t.duration_since_epoch(), 7ns);
  EXPECT_THROW(t -= 1ms, ArithmeticException);

  EXPECT_EQ(t + 5ns, Timestamp::from_duration_since_epoch(12ns));
  EXPECT_THROW(t + -1ms, ArithmeticException);
  EXPECT_EQ(t - 2ns, Timestamp::from_duration_since_epoch(5ns));
  EXPECT_EQ(t - Timestamp::from_duration_since_epoch(3ns), 4ns);
  EXPECT_THROW(t - 1ms, ArithmeticException);

  EXPECT_TRUE(t == Timestamp::from_duration_since_epoch(7ns));
  EXPECT_FALSE(t == Timestamp::from_duration_since_epoch(8ns));

  EXPECT_FALSE(t != Timestamp::from_duration_since_epoch(7ns));
  EXPECT_TRUE(t != Timestamp::from_duration_since_epoch(8ns));

  EXPECT_TRUE(Timestamp::from_duration_since_epoch(2ns) < Timestamp::from_duration_since_epoch(3ns));
  EXPECT_FALSE(Timestamp::from_duration_since_epoch(3ns) < Timestamp::from_duration_since_epoch(2ns));
  EXPECT_FALSE(Timestamp::from_duration_since_epoch(3ns) < Timestamp::from_duration_since_epoch(3ns));

  EXPECT_FALSE(Timestamp::from_duration_since_epoch(2ns) > Timestamp::from_duration_since_epoch(3ns));
  EXPECT_TRUE(Timestamp::from_duration_since_epoch(3ns) > Timestamp::from_duration_since_epoch(2ns));
  EXPECT_FALSE(Timestamp::from_duration_since_epoch(3ns) > Timestamp::from_duration_since_epoch(3ns));

  EXPECT_TRUE(Timestamp::from_duration_since_epoch(2ns) <= Timestamp::from_duration_since_epoch(3ns));
  EXPECT_TRUE(Timestamp::from_duration_since_epoch(3ns) <= Timestamp::from_duration_since_epoch(3ns));
  EXPECT_FALSE(Timestamp::from_duration_since_epoch(3ns) <= Timestamp::from_duration_since_epoch(2ns));

  EXPECT_FALSE(Timestamp::from_duration_since_epoch(2ns) >= Timestamp::from_duration_since_epoch(3ns));
  EXPECT_TRUE(Timestamp::from_duration_since_epoch(3ns) >= Timestamp::from_duration_since_epoch(3ns));
  EXPECT_TRUE(Timestamp::from_duration_since_epoch(3ns) >= Timestamp::from_duration_since_epoch(2ns));
}

File: yogi-dotnet/test/TimestampTest.cs
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

using System;
using Xunit;

namespace test
{
    public class TimestampTest : TestCase
    {
        [Fact]
        public void FromDurationSinceEpoch()
        {
            Assert.Equal(Yogi.Duration.FromDays(123), Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromDays(123)).DurationSinceEpoch);
            Assert.Throws<ArithmeticException>(() =>
            {
                var _ = Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.Inf);
            });
            Assert.Throws<ArithmeticException>(() =>
            {
                var _ = Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(-1));
            });
        }

        [Fact]
        public void Now()
        {
            MOCK_GetCurrentTime((ref long timestamp) =>
            {
                timestamp = 5555555555L;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal(5555555555L, Yogi.Timestamp.Now.DurationSinceEpoch.NanosecondsCount);
        }

        [Fact]
        public void NowError()
        {
            MOCK_GetCurrentTime((ref long timestamp) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.Throws<Yogi.FailureException>(() => Yogi.Timestamp.Now);
        }

        [Fact]
        public void Parse()
        {
            MOCK_ParseTime((ref long timestamp, string str, string timefmt) =>
            {
                Assert.Equal("foo", str);
                Assert.Null(timefmt);
                timestamp = 123456789123000000L;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal(123456789123000000L, Yogi.Timestamp.Parse("foo").DurationSinceEpoch.NanosecondsCount);
        }

        [Fact]
        public void ParseError()
        {
            MOCK_ParseTime((ref long timestamp, string str, string timefmt) =>
            {
                Assert.Equal("foo", str);
                Assert.Equal("bar", timefmt);
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.Throws<Yogi.FailureException>(() => Yogi.Timestamp.Parse("foo", "bar"));
        }

        [Fact]
        public void Constructor()
        {
            Assert.Equal(0, new Yogi.Timestamp().DurationSinceEpoch.NanosecondsCount);
        }

        [Fact]
        public void DurationSinceEpoch()
        {
            var dur = Yogi.Duration.FromDays(123);
            var t = Yogi.Timestamp.FromDurationSinceEpoch(dur);
            Assert.Equal(dur, t.DurationSinceEpoch);
        }

        [Fact]
        public void Nanoseconds()
        {
            Assert.Equal(789, Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(123456789L)).Nanoseconds);
        }

        [Fact]
        public void Microseconds()
        {
            Assert.Equal(456, Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(123456789L)).Microseconds);
        }

        [Fact]
        public void Milliseconds()
        {
            Assert.Equal(123, Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(123456789L)).Milliseconds);
        }

        [Fact]
        public void Format()
        {
            MOCK_FormatTime((long timestamp, ref IntPtr str, ref int strsize, string timefmt) =>
            {
                Assert.Equal(123456789123456789L, timestamp);
                Assert.Null(timefmt);
                str = helloBytes;
                strsize = helloSize;
                return (int)Yogi.ErrorCode.Ok;
            });

            var t = Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(123456789123456789L));
            Assert.Equal(helloString, t.Format());
        }

        [Fact]
        public void FormatError()
        {
            MOCK_FormatTime((long timestamp, ref IntPtr str, ref int strsize, string timefmt) =>
            {
                Assert.Equal("foo", timefmt);
                return (int)Yogi.ErrorCode.Unknown;
            });

            var t = Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(123456789123456789L));
            Assert.Throws<Yogi.FailureException>(() => t.Format("foo"));
        }

        [Fact]
        public void ToStringMethod()
        {
            MOCK_FormatTime((long timestamp, ref IntPtr str, ref int strsize, string timefmt) =>
            {
                Assert.Equal(123456789123456789L, timestamp);
                Assert.Null(timefmt);
                str = helloBytes;
                strsize = helloSize;
                return (int)Yogi.ErrorCode.Ok;
            });

            var t = Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(123456789123456789L));
            Assert.Equal(helloString, t.ToString());
        }

        [Fact]
        public void Operators()
        {
            var t = new Yogi.Timestamp();
            t += Yogi.Duration.FromNanoseconds(8);
            Assert.Equal(8, t.DurationSinceEpoch.NanosecondsCount);
            Assert.Throws<ArithmeticException>(() =>
            {
                t += Yogi.Duration.FromMilliseconds(-1);
            });

            t -= Yogi.Duration.FromNanoseconds(1);
            Assert.Equal(7, t.DurationSinceEpoch.NanosecondsCount);
            Assert.Throws<ArithmeticException>(() =>
            {
                t -= Yogi.Duration.FromMilliseconds(1);
            });

            Assert.Equal(12, (t + Yogi.Duration.FromNanoseconds(5)
                ).DurationSinceEpoch.NanosecondsCount);
            Assert.Throws<ArithmeticException>(() =>
            {
                var _ = t + Yogi.Duration.FromMilliseconds(-1);
            });

            Assert.Equal(5, (t - Yogi.Duration.FromNanoseconds(2)
                ).DurationSinceEpoch.NanosecondsCount);
            Assert.Equal(4, (t - Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(3))).NanosecondsCount);
            Assert.Throws<ArithmeticException>(() =>
            {
                var _ = t - Yogi.Duration.FromMilliseconds(1);
            });

            Assert.True(t == Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(7)));
            Assert.False(t == Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(8)));
            Assert.False(t != Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(7)));
            Assert.True(t != Yogi.Timestamp.FromDurationSinceEpoch(
                Yogi.Duration.FromNanoseconds(8)));

            Assert.True(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2))
                < Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3)));
            Assert.False(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3))
                < Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2)));
            Assert.False(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3))
                < Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3)));

            Assert.False(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2))
                > Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3)));
            Assert.True(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3))
                > Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2)));
            Assert.False(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3))
                > Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3)));

            Assert.True(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2))
                <= Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3)));
            Assert.False(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3))
                <= Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2)));
            Assert.True(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3))
                <= Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3)));

            Assert.False(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2))
                >= Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3)));
            Assert.True(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3))
                >= Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2)));
            Assert.True(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3))
                >= Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(3)));
        }

        [Fact]
        public void EqualsMethod()
        {
            Assert.True(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))
                .Equals(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))));
            Assert.False(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))
                .Equals(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2))));
            Assert.False(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))
                .Equals(""));
        }

        [Fact]
        public void GetHashCodeMethod()
        {
            Assert.NotEqual(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1)),
                Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2)));
        }

        [Fact]
        public void CompareTo()
        {
            Assert.Equal(-1, Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))
                .CompareTo(""));
            Assert.Equal(-1, Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))
                .CompareTo(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2))));
            Assert.Equal(0, Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))
                .CompareTo(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))));
            Assert.Equal(+1, Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(2))
                .CompareTo(Yogi.Timestamp.FromDurationSinceEpoch(Yogi.Duration.FromDays(1))));
        }
    }
}
